File: crucible/ui/application.py
"""
CrucibleApplication — Main Adw.Application subclass.
Handles app lifecycle, CSS loading, actions, and stress-ng cleanup.
"""
from pathlib import Path

import gi
gi.require_version("Gtk", "4.0")
gi.require_version("Adw", "1")
from gi.repository import Gtk, Adw, Gio, Gdk  # noqa: E402

from crucible.version import __version__
from crucible.backend.hardware import gather_specs
from crucible.backend.monitor import SystemMonitor
from crucible.backend.stress import StressManager
from crucible.ui.window import CrucibleWindow


APP_ID = "io.github.sugarycandybar.Crucible"
APP_NAME = "Crucible"


class CrucibleApplication(Adw.Application):
    """Main Crucible application."""

    def __init__(self):
        super().__init__(
            application_id=APP_ID,
            flags=Gio.ApplicationFlags.DEFAULT_FLAGS,
        )
        self._window = None
        self._specs = None
        self._monitor = SystemMonitor()
        self._stress = StressManager()

    def do_startup(self):
        """Load CSS, gather hardware specs, register actions."""
        Adw.Application.do_startup(self)
        self._load_css()
        self._register_packaged_icons()
        self._specs = gather_specs()
        self._setup_actions()

    def do_activate(self):
        """Show the main window."""
        if not self._window:
            self._window = CrucibleWindow(
                specs=self._specs,
                monitor=self._monitor,
                stress=self._stress,
                application=self,
            )
        self._window.present()

    def do_shutdown(self):
        """Kill stress-ng if it's still running."""
        self._stress.kill()
        Adw.Application.do_shutdown(self)

    # --- CSS ------------------------------------------------------------

    def _load_css(self):
        css_provider = Gtk.CssProvider()
        css_path = Path(__file__).parent / "style.css"
        if css_path.exists():
            css_provider.load_from_path(str(css_path))
            Gtk.StyleContext.add_provider_for_display(
                Gdk.Display.get_default(),
                css_provider,
                Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION,
            )

    def _register_packaged_icons(self):
        """Make packaged icons discoverable."""
        display = Gdk.Display.get_default()
        if not display:
            return
            
        icon_theme = Gtk.IconTheme.get_for_display(display)
        
        # Load GResource if installed
        for p in (Path(__file__).resolve().parents[1], Path(__file__).resolve().parents[2]):
            resource_path = p / "crucible-resources.gresource"
            if resource_path.exists():
                resource = Gio.Resource.load(str(resource_path))
                Gio.resources_register(resource)
                icon_theme.add_resource_path("/io/github/sugarycandybar/Crucible/icons")
                break

        # Load from source tree during development
        icon_dir = Path(__file__).resolve().parents[2] / "packaging" / "linux"
        if icon_dir.exists():
            icon_theme = Gtk.IconTheme.get_for_display(display)
            icon_theme.add_search_path(str(icon_dir))

    # --- actions --------------------------------------------------------

    def _setup_actions(self):
        action_about = Gio.SimpleAction.new("about", None)
        action_about.connect("activate", self._on_about)
        self.add_action(action_about)

        action_quit = Gio.SimpleAction.new("quit", None)
        action_quit.connect("activate", self._on_quit)
        self.add_action(action_quit)

        self.set_accels_for_action("app.quit", ["<Primary>q"])

    def _on_about(self, _action, _param):
        dialog = Adw.AboutDialog(
            application_name=APP_NAME,
            application_icon=APP_ID,
            version=__version__,
            developer_name="Sugarycandybar",
            website="https://github.com/sugarycandybar/Crucible",
            issue_url="https://github.com/sugarycandybar/Crucible/issues",
            license_type=Gtk.License.GPL_3_0,
            comments="View hardware specs and stress test your system.",
        )
        dialog.present(self._window)

    def _on_quit(self, _action, _param):
        self.quit()

File: src/application.rs
use std::cell::RefCell;
use std::rc::Rc;

use gtk4::prelude::*;
use libadwaita::prelude::*;

use crate::backend::hardware::gather_specs;
use crate::backend::monitor::SystemMonitor;
use crate::config;
use crate::stability_view;
use crate::window;

fn ensure_schema() {
    let _schema_id = "io.github.sugarycandybar.Crucible.Stability";
    let _already_set = std::env::var("GSETTINGS_SCHEMA_DIR")
        .ok()
        .is_some_and(|d| !d.is_empty());

    // Try to find and compile the schema from the source tree
    let exe_dir = std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|p| p.to_path_buf()));
    let search_paths = vec![
        std::path::PathBuf::from("data"),
        exe_dir
            .as_ref()
            .map(|d| d.join("../../data"))
            .unwrap_or_default(),
        exe_dir
            .as_ref()
            .map(|d| d.join("../../../data"))
            .unwrap_or_default(),
    ];

    for path in &search_paths {
        let schema_path = path.join("io.github.sugarycandybar.Crucible.Stability.gschema.xml");
        if schema_path.exists() {
            let _ = std::process::Command::new("glib-compile-schemas")
                .arg(path)
                .status();
            let abs = std::fs::canonicalize(path).ok();
            if let Some(abs_path) = abs {
                let existing = std::env::var("GSETTINGS_SCHEMA_DIR").unwrap_or_default();
                let combined = if existing.is_empty() {
                    abs_path.to_string_lossy().to_string()
                } else {
                    format!("{}:{}", abs_path.to_string_lossy(), existing)
                };
                std::env::set_var("GSETTINGS_SCHEMA_DIR", &combined);
            }
            break;
        }
    }
}

pub fn run_app() {
    ensure_schema();
    let app = libadwaita::Application::new(
        Some(config::APP_ID),
        gio::ApplicationFlags::default(),
    );
    app.set_resource_base_path(Some("/io/github/sugarycandybar/Crucible"));

    let monitor = Rc::new(RefCell::new(SystemMonitor::new()));
    let stability_state_holder: Rc<RefCell<Option<Rc<RefCell<stability_view::StabilityState>>>>>
        = Rc::new(RefCell::new(None));

    app.connect_startup(|app| {
        // Load CSS
        let provider = gtk4::CssProvider::new();
        provider.load_from_string(include_str!("style.css"));

        if let Some(display) = gtk4::gdk::Display::default() {
            gtk4::style_context_add_provider_for_display(
                &display,
                &provider,
                gtk4::STYLE_PROVIDER_PRIORITY_APPLICATION,
            );
        }

        // Register embedded GResource icons
        let resource_data = include_bytes!(concat!(env!("OUT_DIR"), "/crucible.gresource"));
        let bytes = glib::Bytes::from_static(resource_data);
        if let Ok(resource) = gio::Resource::from_data(&bytes) {
            gio::resources_register(&resource);
            if let Some(display) = gtk4::gdk::Display::default() {
                let icon_theme = gtk4::IconTheme::for_display(&display);
                icon_theme.add_resource_path("/io/github/sugarycandybar/Crucible/icons");
            }
        }

        // Setup actions
        let about_action = gio::SimpleAction::new("about", None);
        let app_weak = app.downgrade();
        about_action.connect_activate(move |_, _| {
            if let Some(app) = app_weak.upgrade() {
                let window = app.active_window();
                let dialog = libadwaita::AboutDialog::builder()
                    .application_name(config::APP_NAME)
                    .application_icon(config::APP_ID)
                    .version(config::VERSION)
                    .developer_name("Sugarycandybar")
                    .website("https://github.com/sugarycandybar/Crucible")
                    .issue_url("https://github.com/sugarycandybar/Crucible/issues")
                    .license_type(gtk4::License::Gpl30)
                    .comments("View hardware specs and stress test your system.")
                    .build();
                dialog.add_acknowledgement_section(
                    Some("Acknowledgements"),
                    &["stress-ng https://github.com/ColinIanKing/stress-ng"],
                );
                dialog.add_other_app(
                    "io.github.sugarycandybar.Hosty",
                    "Hosty",
                    "Host Minecraft servers",
                );
                dialog.add_other_app(
                    "io.github.sugarycandybar.Carabiner",
                    "Carabiner",
                    "Create and manage network tunnels",
                );
                dialog.present(window.as_ref());
            }
        });
        app.add_action(&about_action);

        let quit_action = gio::SimpleAction::new("quit", None);
        let app_weak = app.downgrade();
        quit_action.connect_activate(move |_, _| {
            if let Some(app) = app_weak.upgrade() {
                app.quit();
            }
        });
        app.add_action(&quit_action);
        app.set_accels_for_action("app.quit", &["<Primary>q"]);
    });

    let monitor_clone = Rc::clone(&monitor);
    let holder_clone = Rc::clone(&stability_state_holder);

    app.connect_activate(move |app| {
        if !app.windows().is_empty() {
            app.windows().first().unwrap().present();
            return;
        }

        let specs = gather_specs();

        let (window, stability_state) =
            window::create_window(specs, Rc::clone(&monitor_clone), app);

        *holder_clone.borrow_mut() = Some(Rc::clone(&stability_state));
        window.present();
    });

    app.connect_shutdown(move |_| {
        if let Some(ref state) = *stability_state_holder.borrow() {
            state.borrow_mut().stress.kill();
        }
    });

    app.run();
}
